compiler/rustc_mir_transform/src/gvn.rs

```diff
@@ -1611,6 +1606,21 @@ impl<'tcx> MutVisitor<'tcx> for VnState<'_, 'tcx> {
         }
         self.super_statement(stmt, location);
     }
+
+    fn visit_terminator(&mut self, terminator: &mut Terminator<'tcx>, location: Location) {
+        if let Terminator { kind: TerminatorKind::Call { destination, .. }, .. } = terminator {
+            if let Some(local) = destination.as_local()
+                && self.ssa.is_ssa(local)
+            {
+                let opaque = self.new_opaque();
+                self.assign(local, opaque);
+            }
+            // Function calls maybe invalidate nested deref, and non-local assignments maybe invalidate deref.
+            // Currently, no distinction is made between these two cases.
+            self.invalidate_derefs();
```

Comment:
I fear we are missing many cases of indirect writes. For example, inline asm.

Reply:
Instead of addressing to which terminator may write something, I adopted a more conservative approach, which terminator must not write anything.